SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorPackageAssembly.cpp

```diff
@@ -0,0 +1,426 @@
+}
+
+static void WriteSimulatedPlayerCoordinatorShellScript()
+{
+	FString ShellScript = FString::Printf(TEXT(
+		"#!/bin/sh\n"
+		"\n"
+		"# Some clients are quite large so in order to avoid running out of disk space on the node we attempt to delete the zip\n"
+		"WORKER_ZIP_DIR=\"/tmp/runner_source/\"\n"
+		"if [ -d \"$WORKER_ZIP_DIR\" ]; then\n"
+		"  rm -rf \"$WORKER_ZIP_DIR\"\n"
+		"fi\n"
+		"\n"
+		"sleep 5\n"
+		"\n"
+		"chmod +x WorkerCoordinator.exe\n"
+		"chmod +x StartSimulatedClient.sh\n"
+		"chmod +x %s.sh\n"
+		"\n"
+		"mono WorkerCoordinator.exe $@ 2> /improbable/logs/CoordinatorErrors.log\n"
+		), FApp::GetProjectName());
+
+	FString StagingDir = GetStagingDir();
+	FString OutputFile = StagingDir / TEXT("LinuxNoEditor") / TEXT("StartCoordinator.sh");
+	FFileHelper::SaveStringToFile(ShellScript, *OutputFile, FFileHelper::EEncodingOptions::ForceAnsi);
+}
+
+static void ZipWorker(const FString& WorkerName, const FString& ZipName, TFunction<void(FString,double)> Func)
+{
+	//write shell script and zip folder
+	FString ProjectPath = FPaths::IsProjectFilePathSet() ? FPaths::ConvertRelativePathToFull(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName() / FApp::GetProjectName() + TEXT(".uproject");
+
+	FString SourcePath = FPaths::ConvertRelativePathToFull((FPaths::IsProjectFilePathSet() ? FPaths::GetPath(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName()) / TEXT("..") / TEXT("spatial") / TEXT("build") / TEXT("unreal") / WorkerName);
+	FString AssemblyPath = FPaths::ConvertRelativePathToFull((FPaths::IsProjectFilePathSet() ? FPaths::GetPath(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName()) / TEXT("..") / TEXT("spatial") / TEXT("build") / TEXT("assembly") / TEXT("worker") / ZipName);
+	FString CommandLine = FString::Printf(TEXT("-ScriptsForProject=\"%s\" ZipUtils -add=\"%s\" -archive=\"%s\""),
+		*ProjectPath, *SourcePath, *AssemblyPath);
+
+	AsyncTask(ENamedThreads::GameThread, [CommandLine, Func]() {
+		IUATHelperModule::Get().CreateUatTask(CommandLine,
+			LOCTEXT("ZipAssemblyDisplayName", "Spatial Cloud"),
+			LOCTEXT("ZipAssemblyDescription", "Zip Cloud Deployment Assembly"),
+			LOCTEXT("ZipAssemblyShortName", "Cloud Assembly"),
+			FEditorStyle::GetBrush(TEXT("MainFrame.CookContent")),
+			TFunction<void(FString, double)>(Func)
+			);
+		});
+}
+
+static void Build(const FString& CommandLine, TFunction<void(FString, double)> Func)
+{
+	AsyncTask(ENamedThreads::GameThread, [CommandLine, Func]() {
+		IUATHelperModule::Get().CreateUatTask(CommandLine,
+			LOCTEXT("BuildAssemblyDisplayName", "Spatial Cloud"),
+			LOCTEXT("BuildAssemblyDescription", "Build Cloud Deployment Assembly"),
+			LOCTEXT("BuildAssemblyShortName", "Cloud Assembly"),
+			FEditorStyle::GetBrush(TEXT("MainFrame.CookContent")),
+			TFunction<void(FString, double)>(Func)
+			);
+		});
+}
+
+static FString GetServerBuildCommand(const FString& OptionalParams)
```

Comment:
For building I think we should have an integration with UBT directly since we are already in the editor. We can make use of Unreal's packaging tool (in the editor) rather than calling UBT via a new process. I believe this will also give us the ability to use preexisting commandlets, although that last bit needs investigating

Reply:
This code was modelled after the File->Package Project code, since we do a full packaging, I reviewed how that option worked in the editor and followed the same pattern as Unreal...? I'm not really sure I understand this comment, separate processes provide protection through isolation, is there any reason a separate process is undesirable?